src/main/scala/uk/gov/nationalarchives/omega/api/services/ApiService.scala

```diff
@@ -0,0 +1,219 @@
+    val messageStoreFolder = Paths.get(config.tempMessageDir)
+    Files.createDirectories(messageStoreFolder)
+    new LocalMessageStore(messageStoreFolder)
+  }
+
+  def start: IO[ExitCode] = {
+    // check if we can start the service
+    if (!switchState(Stopped, Starting)(config.errorOnInvalidServiceState)) {
+      return IO.pure(ExitCode.Error)
+    }
+
+    // copy config option to local state
+    this.errorOnInvalidServiceState = config.errorOnInvalidServiceState
+
+    try {
+      // attempt startup
+      val ec = doStart(config)
+      // switch to the started state
+      switchState(Starting, Started)(errorOnInvalidServiceState)
+      ec
+    } catch {
+      case e: Throwable =>
+        // we are in some inconsistent state as startup failed, so we must attempt shutdown
+        try {
+          doStop()
+          switchState(Starting, Stopped)(errorOnInvalidServiceState)
+          IO.pure(ExitCode.Error)
+        } finally
+          println(e.getMessage)
+    }
+
+  }
+
+  @throws[IllegalStateException]
+  def stop(): Unit = {
+    // check if we can stop the service
+    if (!switchState(Started, Stopping)(errorOnInvalidServiceState)) {
+      return
+    }
+
+    // attempt shutdown
+    doStop()
+    switchState(Stopping, Stopped)(errorOnInvalidServiceState)
+    () // Explicitly return unit
+  }
+
+  @throws[IllegalStateException]
+  private def switchState(from: ServiceState, to: ServiceState)(errorOnInvalidServiceState: Boolean): Boolean = {
+    val switched = state.compareAndSet(from, to)
+    if (!switched) {
+      val msg = s"Unable to switch ApiService state from: $from, to: $to."
+      if (errorOnInvalidServiceState) {
+        throw new IllegalStateException(msg)
+      } else {
+        logger.warn(msg)
+      }
+    }
+    switched
+  }
+
+  private def doStart(config: ServiceConfig): IO[ExitCode] = {
+
+    // TODO(AR) - one client, how to ack a consumer message after local persistence and then process it, and then produce a response
+    // TODO(AR) how to wire up queues and services using a config file or DSL?
+
+    // TODO(AR) request queue will typically be 1 (plus maybe a few more for expedited ops), response queues will be per external application
+
+    val localQueue: IO[Queue[IO, LocalMessage]] = Queue.bounded[IO, LocalMessage](config.maxLocalQueueSize)
+    val jmsConnector = new JmsConnector(config)
+
+    val result = for {
+      q <- localQueue
+      res <- jmsConnector.getJmsIo(QueueName(config.requestQueue)).use { case (jmsProducer, jmsConsumer) =>
+               process(q, jmsConsumer, jmsProducer)
+             }
+    } yield res
+    result.as(ExitCode.Success)
+  }
+
+  private def process(q: Queue[IO, LocalMessage], consumer: JmsAcknowledgerConsumer[IO], producer: JmsProducer[IO]) =
+    for {
+      // START
+      res <- IO.race(
+               transferMessageToLocalQueue(q)(consumer),
+
+               // single dispatcher
+               //        dispatcher(q).foreverM
```

Comment:
Is this a comment or a commented out statement?

Reply:
Removed.

Reply:
Confirmed.